From d97d1d7651ad5b3fc5411145942f06d65be9bcce Mon Sep 17 00:00:00 2001
From: MartinChoo <214582617@qq.com>
Date: Wed, 23 Jul 2025 17:43:08 +0800
Subject: [PATCH 07/12] DatabaseDebugTool

---
 src/shell.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/shell.c b/src/shell.c
index 7fb190e..c1948b1 100644
--- a/src/shell.c
+++ b/src/shell.c
@@ -119,6 +119,7 @@ typedef unsigned short int u16;
 #include <assert.h>
 #include <math.h>
 #include "sqlite3.h"
+#include "sqlite3sym.h"
 typedef sqlite3_int64 i64;
 typedef sqlite3_uint64 u64;
 typedef unsigned char u8;
@@ -28875,6 +28876,9 @@ int SQLITE_CDECL wmain(int argc, wchar_t **wargv){
   }
 #endif
 
+// The icu analyzer is enabled by default on the client tool.
+  sqlite3_config(SQLITE_CONFIG_ENABLE_ICU);
+
   /* Do an initial pass through the command-line argument to locate
   ** the name of the database file, the name of the initialization file,
   ** the size of the alternative malloc heap, options affecting commands
-- 
2.47.0.windows.2

